File: src/Libraries/Security/apple/OSX/sec/Security/SecSharedCredential.m
/*
 * Copyright (c) 2020 Apple Inc. All Rights Reserved.
 *
 * @APPLE_LICENSE_HEADER_START@
 *
 * This file contains Original Code and/or Modifications of Original Code
 * as defined in and that are subject to the Apple Public Source License
 * Version 2.0 (the 'License'). You may not use this file except in
 * compliance with the License. Please obtain a copy of the License at
 * http://www.opensource.apple.com/apsl/ and read it before using this
 * file.
 *
 * The Original Code and all software distributed under the License are
 * distributed on an 'AS IS' basis, WITHOUT WARRANTY OF ANY KIND, EITHER
 * EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
 * INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, QUIET ENJOYMENT OR NON-INFRINGEMENT.
 * Please see the License for the specific language governing rights and
 * limitations under the License.
 *
 * @APPLE_LICENSE_HEADER_END@
 *
 * SecSharedCredential.m - Retrieve shared credentials with AuthenticationServices.
 *
 */

#include <Security/SecSharedCredential.h>
#include <Security/SecBasePriv.h>
#include <utilities/SecCFError.h>
#include <utilities/SecCFWrappers.h>
#include "SecItemInternal.h"
#include <dlfcn.h>

#import <Foundation/Foundation.h>
#import <AuthenticationServices/AuthenticationServices.h>

// Forward declaration of the primary function implemented in this file
OSStatus SecCopySharedWebCredentialSyncUsingAuthSvcs(CFStringRef fqdn, CFStringRef account, CFArrayRef *credentials, CFErrorRef *error);

// Classes we will load dynamically
static Class kASAuthorizationClass = NULL;
static Class kASAuthorizationControllerClass = NULL;
static Class kASAuthorizationPasswordProviderClass = NULL;
static Class kASPasswordCredentialClass = NULL;
static Class kUIApplicationClass = NULL;
static Class kNSApplicationClass = NULL;

static void loadAuthenticationServices(void) {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        const char *path = "/System/Library/Frameworks/AuthenticationServices.framework/AuthenticationServices";
        if ( [NSProcessInfo processInfo].macCatalystApp == YES ) {
            path = "/System/iOSSupport/System/Library/Frameworks/AuthenticationServices.framework/AuthenticationServices";
        }
        void* lib_handle = dlopen(path, RTLD_LAZY);
        if (lib_handle != NULL) {
            kASAuthorizationClass = NSClassFromString(@"ASAuthorization");
            kASAuthorizationControllerClass = NSClassFromString(@"ASAuthorizationController");
            kASAuthorizationPasswordProviderClass = NSClassFromString(@"ASAuthorizationPasswordProvider");
            kASPasswordCredentialClass = NSClassFromString(@"ASPasswordCredential");
        }
    });
}

static void loadUIKit(void) {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        const char *path = "/System/Library/Frameworks/UIKit.framework/UIKit";
        if ( [NSProcessInfo processInfo].macCatalystApp == YES ) {
            path = "/System/Library/iOSSupport/System/Library/Frameworks/UIKit.framework/UIKit";
        }
        void* lib_handle = dlopen(path, RTLD_LAZY);
        if (lib_handle != NULL) {
            kUIApplicationClass = NSClassFromString(@"UIApplication");
        }
    });
}

static void loadAppKit(void) {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        const char *path = "/System/Library/Frameworks/AppKit.framework/AppKit";
        void* lib_handle = dlopen(path, RTLD_LAZY);
        if (lib_handle != NULL) {
            kNSApplicationClass = NSClassFromString(@"NSApplication");
        }
    });
}

static Class ASAuthorizationClass() {
    loadAuthenticationServices();
    return kASAuthorizationClass;
}

static Class ASAuthorizationControllerClass() {
    loadAuthenticationServices();
    return kASAuthorizationControllerClass;
}

static Class ASAuthorizationPasswordProviderClass() {
    loadAuthenticationServices();
    return kASAuthorizationPasswordProviderClass;
}

static Class ASPasswordCredentialClass() {
    loadAuthenticationServices();
    return kASPasswordCredentialClass;
}

static Class UIApplicationClass() {
    loadUIKit();
    return kUIApplicationClass;
}

static Class NSApplicationClass() {
    loadAppKit();
    return kNSApplicationClass;
}

@interface SharedCredentialController : NSObject
    <ASAuthorizationControllerDelegate,
     ASAuthorizationControllerPresentationContextProviding>

-(ASPasswordCredential *)passwordCredential;

@end

@implementation SharedCredentialController {
    ASAuthorizationPasswordProvider *_provider;
    ASAuthorizationController *_controller;
    ASPasswordCredential *_passwordCredential;
    dispatch_semaphore_t _semaphore;
    NSError *_error;
    OSStatus _result;
}

- (void)dealloc {
    // Don't want any further callbacks since we are going away
    _controller.delegate = nil;
    _controller.presentationContextProvider = nil;
}

- (void)_requestCredential {
    if (!_provider) {
        _provider = [[ASAuthorizationPasswordProviderClass() alloc] init];
    }
    if (!_controller) {
        _controller = [[ASAuthorizationControllerClass() alloc] initWithAuthorizationRequests:@[ [_provider createRequest] ]];
    }
    _controller.delegate = self;
    _controller.presentationContextProvider = self;
    _semaphore = dispatch_semaphore_create(0);
    _result = errSecItemNotFound;
    _error = nil;

    [_controller performRequests];
}

- (ASPasswordCredential *)passwordCredential {
    if (_passwordCredential) {
        return _passwordCredential;
    }
    BOOL shouldRequest = YES; // ( [NSProcessInfo processInfo].macCatalystApp == YES );
    if (shouldRequest) {
        [self _requestCredential];
        // wait synchronously until user picks a credential or cancels
        dispatch_semaphore_wait(_semaphore, DISPATCH_TIME_FOREVER);
    } else {
        // unable to return a shared credential: <rdar://problem/59958701>
        _result = errSecItemNotFound;
        _error = [[NSError alloc] initWithDomain:NSOSStatusErrorDomain code:_result userInfo:NULL];
    }
    return _passwordCredential;
}

- (NSError *)error {
    return _error;
}

- (OSStatus)result {
    return _result;
}

- (void)authorizationController:(ASAuthorizationController *)controller didCompleteWithAuthorization:(ASAuthorization *)authorization {
    secinfo("swcagent", "SWC received didCompleteWithAuthorization");
    ASPasswordCredential *passwordCredential = authorization.credential;
    if (![passwordCredential isKindOfClass:[ASPasswordCredentialClass() class]]) {
        _passwordCredential = nil;
        _result = errSecItemNotFound;
    } else {
        _passwordCredential = passwordCredential;
        _result = errSecSuccess;
    }
    dispatch_semaphore_signal(_semaphore);
}

- (void)authorizationController:(ASAuthorizationController *)controller didCompleteWithError:(NSError *)error {
    secinfo("swcagent", "SWC received didCompleteWithError");
    _passwordCredential = nil;
    _error = error;
    _result = errSecItemNotFound;
    dispatch_semaphore_signal(_semaphore);
}

- (ASPresentationAnchor)presentationAnchorForAuthorizationController:(ASAuthorizationController *)controller
{
    ASPresentationAnchor anchorWindow = nil;
#if TARGET_OS_OSX
    if ( [NSProcessInfo processInfo].macCatalystApp == NO ) {
        anchorWindow = [[NSApplicationClass() sharedApplication] keyWindow];
    }
#endif
    if (!anchorWindow) {
        anchorWindow = [[UIApplicationClass() sharedApplication] keyWindow];
    }
    return anchorWindow;
}

@end

OSStatus SecCopySharedWebCredentialSyncUsingAuthSvcs(CFStringRef fqdn, CFStringRef account, CFArrayRef *credentials, CFErrorRef *error) {
    SharedCredentialController *controller = [[SharedCredentialController alloc] init];
    ASPasswordCredential *passwordCredential = [controller passwordCredential];
    OSStatus status = [controller result];
    NSArray *returnedCredentials = @[];
    if (status != errSecSuccess) {
        secinfo("swcagent", "SecCopySharedWebCredentialSyncUsingAuthSvcs received result %d", (int)status);
        if (error) {
            *error = (CFErrorRef)CFBridgingRetain([controller error]);
        }
    } else if (passwordCredential) {
        // Use the .user and .password of the passwordCredential to satisfy the SWC interface.
        NSDictionary *credential = @{
            (id)kSecAttrServer : (__bridge NSString*)fqdn,
            (id)kSecAttrAccount : passwordCredential.user,
            (id)kSecSharedPassword : passwordCredential.password,
        };
        returnedCredentials = @[ credential ];
    } else {
        secinfo("swcagent", "SecCopySharedWebCredentialSyncUsingAuthSvcs found no credential");
        status = errSecItemNotFound;
    }
    if (credentials) {
        *credentials = (CFArrayRef)CFBridgingRetain(returnedCredentials);
    }
    return status;
}

File: src/Libraries/Security/apple/OSX/sec/Security/SecSharedCredential.c
/*
 * Copyright (c) 2014-2020 Apple Inc. All Rights Reserved.
 *
 * @APPLE_LICENSE_HEADER_START@
 *
 * This file contains Original Code and/or Modifications of Original Code
 * as defined in and that are subject to the Apple Public Source License
 * Version 2.0 (the 'License'). You may not use this file except in
 * compliance with the License. Please obtain a copy of the License at
 * http://www.opensource.apple.com/apsl/ and read it before using this
 * file.
 *
 * The Original Code and all software distributed under the License are
 * distributed on an 'AS IS' basis, WITHOUT WARRANTY OF ANY KIND, EITHER
 * EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
 * INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, QUIET ENJOYMENT OR NON-INFRINGEMENT.
 * Please see the License for the specific language governing rights and
 * limitations under the License.
 *
 * @APPLE_LICENSE_HEADER_END@
 *
 * SecSharedCredential.c - CoreFoundation-based functions to store and retrieve shared credentials.
 *
 */


#include <Security/SecSharedCredential.h>
#include <Security/SecBasePriv.h>
#include <utilities/SecCFError.h>
#include <utilities/SecCFWrappers.h>
#include "SecItemInternal.h"
#include <ipc/securityd_client.h>
#include "SecPasswordGenerate.h"

/* forward declarations */
OSStatus SecAddSharedWebCredentialSync(CFStringRef fqdn, CFStringRef account, CFStringRef password, CFErrorRef *error);
OSStatus SecCopySharedWebCredentialSync(CFStringRef fqdn, CFStringRef account, CFArrayRef *credentials, CFErrorRef *error);
#if TARGET_OS_OSX || TARGET_OS_MACCATALYST
OSStatus SecCopySharedWebCredentialSyncUsingAuthSvcs(CFStringRef fqdn, CFStringRef account, CFArrayRef *credentials, CFErrorRef *error);
#endif

#if SHAREDWEBCREDENTIALS

// OSX now has SWC enabled, but cannot link SharedWebCredentials framework: rdar://59958701
#if TARGET_OS_OSX || TARGET_OS_MACCATALYST

OSStatus SecAddSharedWebCredentialSync(CFStringRef fqdn,
    CFStringRef account,
    CFStringRef password,
    CFErrorRef *error)
{
    OSStatus status = errSecUnimplemented;
    if (error) {
        SecError(status, error, CFSTR("SecAddSharedWebCredentialSync not supported on this platform"));
    }
    return status;
}

#else

OSStatus SecAddSharedWebCredentialSync(CFStringRef fqdn,
    CFStringRef account,
    CFStringRef password,
    CFErrorRef *error)
{
    OSStatus status;
    __block CFErrorRef* outError = error;
    __block CFMutableDictionaryRef args = CFDictionaryCreateMutable(kCFAllocatorDefault,
        0, &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);
    if (fqdn) {
        CFDictionaryAddValue(args, kSecAttrServer, fqdn);
    }
    if (account) {
        CFDictionaryAddValue(args, kSecAttrAccount, account);
    }
    if (password) {
        CFDictionaryAddValue(args, kSecSharedPassword, password);
    }
    status = SecOSStatusWith(^bool (CFErrorRef *error) {
        CFTypeRef raw_result = NULL;
        bool xpc_result = false;
        bool internal_spi = false; // TODO: support this for SecurityDevTests
        if(internal_spi && gSecurityd && gSecurityd->sec_add_shared_web_credential) {
            xpc_result = gSecurityd->sec_add_shared_web_credential(args, NULL, NULL, NULL, SecAccessGroupsGetCurrent(), &raw_result, error);
        } else {
            xpc_result = cftype_client_to_bool_cftype_error_request(sec_add_shared_web_credential_id, args, SecSecurityClientGet(), &raw_result, error);
        }
        CFReleaseSafe(args);
        if (!xpc_result) {
            if (NULL == *error) {
                SecError(errSecInternal, error, CFSTR("Internal error (XPC failure)"));
            }
        }
        if (outError) {
            *outError = (error) ? *error : NULL;
            CFRetainSafe(*outError);
        } else {
            CFReleaseNull(*error);
        }
        CFReleaseNull(raw_result);
        return xpc_result;
    });

    return status;
}
#endif /* !TARGET_OS_OSX || !TARGET_OS_MACCATALYST */
#endif /* SHAREDWEBCREDENTIALS */

void SecAddSharedWebCredential(CFStringRef fqdn,
    CFStringRef account,
    CFStringRef password,
    void (^completionHandler)(CFErrorRef error))
{
	__block CFErrorRef error = NULL;
	__block dispatch_queue_t dst_queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT,0);
#if SHAREDWEBCREDENTIALS

    /* sanity check input arguments */
	CFStringRef errStr = NULL;
	if (!fqdn || CFGetTypeID(fqdn) != CFStringGetTypeID() || !CFStringGetLength(fqdn) ||
		!account || CFGetTypeID(account) != CFStringGetTypeID() || !CFStringGetLength(account) ) {
		errStr = CFSTR("fqdn or account was not of type CFString, or not provided");
	}
	else if (password && CFGetTypeID(password) != CFStringGetTypeID()) {
		errStr = CFSTR("non-nil password was not of type CFString");
	}
	if (errStr) {
		SecError(errSecParam, &error, CFSTR("%@"), errStr);
		dispatch_async(dst_queue, ^{
			if (completionHandler) {
				completionHandler(error);
			}
			CFReleaseSafe(error);
		});
		return;
	}

	__block CFStringRef serverStr = CFRetainSafe(fqdn);
	__block CFStringRef accountStr = CFRetainSafe(account);
	__block CFStringRef passwordStr = CFRetainSafe(password);

	dispatch_async(dst_queue, ^{
		OSStatus status = SecAddSharedWebCredentialSync(serverStr, accountStr, passwordStr, &error);
		CFReleaseSafe(serverStr);
		CFReleaseSafe(accountStr);
		CFReleaseSafe(passwordStr);

		if (status && !error) {
			SecError(status, &error, CFSTR("Error adding shared password"));
		}
		dispatch_async(dst_queue, ^{
			if (completionHandler) {
				completionHandler(error);
			}
			CFReleaseSafe(error);
		});
	});
#else
    SecError(errSecParam, &error, CFSTR("SharedWebCredentials not supported on this platform"));
    dispatch_async(dst_queue, ^{
        if (completionHandler) {
            completionHandler(error);
        }
        CFReleaseSafe(error);
    });
#endif
}

#if SHAREDWEBCREDENTIALS

#if TARGET_OS_OSX || TARGET_OS_MACCATALYST

OSStatus SecCopySharedWebCredentialSync(CFStringRef fqdn,
    CFStringRef account,
    CFArrayRef *credentials,
    CFErrorRef *error)
{
    OSStatus status = errSecUnimplemented;
    if (error) {
        SecError(status, error, CFSTR("SecCopySharedWebCredentialSync not supported on this platform"));
    }
    return status;
}

#else

OSStatus SecCopySharedWebCredentialSync(CFStringRef fqdn,
    CFStringRef account,
    CFArrayRef *credentials,
    CFErrorRef *error)
{
    OSStatus status;
    __block CFErrorRef* outError = error;
    __block CFMutableDictionaryRef args = CFDictionaryCreateMutable(kCFAllocatorDefault,
        0, &kCFTypeDictionaryKeyCallBacks, &kCFTypeDictionaryValueCallBacks);
    if (fqdn) {
        CFDictionaryAddValue(args, kSecAttrServer, fqdn);
    }
    if (account) {
        CFDictionaryAddValue(args, kSecAttrAccount, account);
    }
    status = SecOSStatusWith(^bool (CFErrorRef *error) {
        CFTypeRef raw_result = NULL;
        bool xpc_result = false;
        bool internal_spi = false; // TODO: support this for SecurityDevTests
        if(internal_spi && gSecurityd && gSecurityd->sec_copy_shared_web_credential) {
            xpc_result = gSecurityd->sec_copy_shared_web_credential(args, NULL, NULL, NULL, SecAccessGroupsGetCurrent(), &raw_result, error);
        } else {
            xpc_result = cftype_client_to_bool_cftype_error_request(sec_copy_shared_web_credential_id, args, SecSecurityClientGet(), &raw_result, error);
        }
        CFReleaseSafe(args);
        if (!xpc_result) {
            if (NULL == *error) {
                SecError(errSecInternal, error, CFSTR("Internal error (XPC failure)"));
            }
        }
        if (outError) {
            *outError = (error) ? *error : NULL;
            CFRetainSafe(*outError);
        } else {
            CFReleaseNull(*error);
        }
        if (!raw_result) {
            raw_result = CFArrayCreateMutable(kCFAllocatorDefault, 0, &kCFTypeArrayCallBacks);
        }
        *credentials = raw_result;
        return xpc_result;
    });

    return status;
}
#endif /* !TARGET_OS_OSX || !TARGET_OS_MACCATALYST */
#endif /* SHAREDWEBCREDENTIALS */

void SecRequestSharedWebCredential(CFStringRef fqdn,
    CFStringRef account,
    void (^completionHandler)(CFArrayRef credentials, CFErrorRef error))
{
	__block CFErrorRef error = NULL;
	__block dispatch_queue_t dst_queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT,0);
#if SHAREDWEBCREDENTIALS
    __block CFArrayRef result = NULL;

    /* sanity check input arguments, if provided */
	CFStringRef errStr = NULL;
	if (fqdn && (CFGetTypeID(fqdn) != CFStringGetTypeID() || !CFStringGetLength(fqdn))) {
		errStr = CFSTR("fqdn was empty or not a CFString");
    }
    else if (account && (CFGetTypeID(account) != CFStringGetTypeID() || !CFStringGetLength(account))) {
		errStr = CFSTR("account was empty or not a CFString");
	}
	if (errStr) {
		SecError(errSecParam, &error, CFSTR("%@"), errStr);
		dispatch_async(dst_queue, ^{
			if (completionHandler) {
				completionHandler(result, error);
			}
			CFReleaseSafe(error);
            CFReleaseSafe(result);
		});
		return;
	}

    __block CFStringRef serverStr = CFRetainSafe(fqdn);
	__block CFStringRef accountStr = CFRetainSafe(account);

    dispatch_async(dst_queue, ^{
#if TARGET_OS_OSX || TARGET_OS_MACCATALYST
		OSStatus status = SecCopySharedWebCredentialSyncUsingAuthSvcs(serverStr, accountStr, &result, &error);
#else
		OSStatus status = SecCopySharedWebCredentialSync(serverStr, accountStr, &result, &error);
#endif
		CFReleaseSafe(serverStr);
		CFReleaseSafe(accountStr);

		if (status && !error) {
			SecError(status, &error, CFSTR("Error copying shared password"));
		}
		dispatch_async(dst_queue, ^{
			if (completionHandler) {
				completionHandler(result, error);
			}
			CFReleaseSafe(error);
			CFReleaseSafe(result);
		});
	});
#else
    SecError(errSecParam, &error, CFSTR("SharedWebCredentials not supported on this platform"));
    dispatch_async(dst_queue, ^{
        if (completionHandler) {
            completionHandler(NULL, error);
        }
        CFReleaseSafe(error);
    });
#endif /* SHAREDWEBCREDENTIALS */

}

CFStringRef SecCreateSharedWebCredentialPassword(void)
{

    CFStringRef password = NULL;
    CFErrorRef error = NULL;
    CFMutableDictionaryRef passwordRequirements = NULL;

    CFStringRef allowedCharacters = CFSTR("abcdefghkmnopqrstuvwxyzABCDEFGHJKLMNPQRSTUVWXYZ23456789");
    CFCharacterSetRef requiredCharactersLower = CFCharacterSetCreateWithCharactersInString(NULL, CFSTR("abcdefghkmnopqrstuvwxyz"));
    CFCharacterSetRef requiredCharactersUppder = CFCharacterSetCreateWithCharactersInString(NULL, CFSTR("ABCDEFGHJKLMNPQRSTUVWXYZ"));
    CFCharacterSetRef requiredCharactersNumbers = CFCharacterSetCreateWithCharactersInString(NULL, CFSTR("3456789"));

    int groupSize = 3;
    int groupCount = 4;
    int totalLength = (groupSize * groupCount);
    CFNumberRef groupSizeRef = CFNumberCreate(NULL, kCFNumberIntType, &groupSize);
    CFNumberRef groupCountRef = CFNumberCreate(NULL, kCFNumberIntType, &groupCount);
    CFNumberRef totalLengthRef = CFNumberCreate(NULL, kCFNumberIntType, &totalLength);
    CFStringRef separator = CFSTR("-");

    CFMutableArrayRef requiredCharacterSets = CFArrayCreateMutable(NULL, 0, &kCFTypeArrayCallBacks);
    CFArrayAppendValue(requiredCharacterSets, requiredCharactersLower);
    CFArrayAppendValue(requiredCharacterSets, requiredCharactersUppder);
    CFArrayAppendValue(requiredCharacterSets, requiredCharactersNumbers);

    passwordRequirements = CFDictionaryCreateMutable(NULL, 0, NULL, NULL);
    CFDictionaryAddValue(passwordRequirements, kSecPasswordAllowedCharactersKey, allowedCharacters);
    CFDictionaryAddValue(passwordRequirements, kSecPasswordRequiredCharactersKey, requiredCharacterSets);
    CFDictionaryAddValue(passwordRequirements, kSecPasswordGroupSize, groupSizeRef );
    CFDictionaryAddValue(passwordRequirements, kSecPasswordNumberOfGroups, groupCountRef);
    CFDictionaryAddValue(passwordRequirements, kSecPasswordSeparator, separator);
    CFDictionaryAddValue(passwordRequirements, kSecPasswordMaxLengthKey, totalLengthRef);
    CFDictionaryAddValue(passwordRequirements, kSecPasswordMinLengthKey, totalLengthRef);
    CFDictionaryAddValue(passwordRequirements, kSecPasswordDefaultForType, CFSTR("false"));
    CFRelease(requiredCharactersLower);
    CFRelease(requiredCharactersUppder);
    CFRelease(requiredCharactersNumbers);
    CFRelease(groupSizeRef);
    CFRelease(groupCountRef);
    CFRelease(totalLengthRef);

    password = SecPasswordGenerate(kSecPasswordTypeSafari, &error, passwordRequirements);

    CFRelease(requiredCharacterSets);
    CFRelease(passwordRequirements);
    if ((error && error != errSecSuccess) || !password)
    {
        if (password) CFRelease(password);
        secwarning("SecPasswordGenerate failed to generate a password for SecCreateSharedWebCredentialPassword.");
        return NULL;
    } else {
        return password;
    }

}
